File: CyclicBFS.java
import java.util.*;

public class CyclicBFS {
    static boolean bfs(int start, List<List<Integer>> graph, boolean[] visited) {
        Queue<Integer> queue = new LinkedList<>();
        int[] parent = new int[graph.size()];
        Arrays.fill(parent, -1);

        visited[start] = true;
        queue.add(start);

        while (!queue.isEmpty()) {
            int node = queue.poll();

            for (int neighbor : graph.get(node)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    parent[neighbor] = node;
                    queue.add(neighbor);
                } else if (parent[node] != neighbor) {
                    System.out.println("Cycle found between: " + node + " and " + neighbor);
                    return true;
                }
            }
        }

        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of vertices: ");
        int n = sc.nextInt();

        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());

        System.out.println("Enter adjacency matrix:");
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                if (sc.nextInt() == 1)
                    graph.get(i).add(j);

        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++)
            if (!visited[i] && bfs(i, graph, visited))
                return;

        System.out.println("No cycle found.");
    }
}

File: Algo&Comp/cyclic_bfs.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

bool bfsCycle(int start, vector<vector<int>>& graph, vector<bool>& visited) {
    vector<int> parent(graph.size(), -1);
    queue<int> q;

    visited[start] = true;
    q.push(start);

    while (!q.empty()) {
        int node = q.front();
        q.pop();

        for (int neighbor = 0; neighbor < graph[node].size(); ++neighbor) {
            if (graph[node][neighbor]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    parent[neighbor] = node;
                    q.push(neighbor);
                } else if (parent[node] != neighbor) {
                    cout << "Cycle found involving: " << node << " and " << neighbor << endl;
                    return true;
                }
            }
        }
    }

    return false;
}

int main() {
    int n;
    cout << "Enter number of vertices: ";
    cin >> n;

    vector<vector<int>> graph(n, vector<int>(n));
    cout << "Enter adjacency matrix:\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> graph[i][j];

    vector<bool> visited(n, false);
    for (int i = 0; i < n; ++i) {
        if (!visited[i] && bfsCycle(i, graph, visited)) {
            return 0;
        }
    }

    cout << "No cycle found.\n";
    return 0;
}
